packages/zql/src/query/named.test.ts

```diff
@@ -27,64 +29,141 @@ test('defining a named query', () => {
+        readonly id: string;
+        readonly title: string;
+        readonly description: string;
+        readonly closed: boolean;
+        readonly ownerId: string | null;
+        readonly createdAt: number;
+      }[]
+    >
   >();
-  expectTypeOf<ReturnType<typeof q2.run>>().toEqualTypeOf<
-    ReturnType<typeof q.run>
+
+  expect(q.customQueryID).toEqual({
+    name: 'myQuery',
+    args: ['123'],
+  });
+
+  // no validator was defined
+  expect(() => withValidation(def)).toThrowErrorMatchingInlineSnapshot(
+    `[Error: ret does not have a validator defined]`,
+  );
+});
+
+test('defining a synced query with validation', () => {
+  const def = syncedQuery(
+    'myQuery',
+    (id: unknown) => {
+      assert(typeof id === 'string', 'id must be a string');
+      return [id] as const;
+    },
+    (id: string) => builder.issue.where('id', id),
+  );
+
+  let q = def('123');
+  expectTypeOf<ReturnType<typeof q.run>>().toEqualTypeOf<
+    Promise<
+      {
+        readonly id: string;
+        readonly title: string;
+        readonly description: string;
+        readonly closed: boolean;
+        readonly ownerId: string | null;
+        readonly createdAt: number;
+      }[]
+    >
   >();
-  expectTypeOf<ReturnType<typeof q3.run>>().toEqualTypeOf<
-    ReturnType<typeof q.run>
+
+  expect(q.customQueryID).toEqual({
+    name: 'myQuery',
+    args: ['123'],
+  });
+
+  const {validator} = def;
+  expectTypeOf<ReturnType<typeof validator>>().toEqualTypeOf<
+    readonly [string] | [string]
   >();
-});
 
-function check(
-  named: NamedQuery<[string], any>,
-  expectedName: string = 'myName',
-) {
-  const r = named('123');
+  const validated = withValidation(def);
+  q = validated('321');
+  expect(q.customQueryID).toEqual({
+    name: 'myQuery',
+    args: ['321'],
+  });
 
-  const id = r.customQueryID;
-  expect(id?.name).toBe(expectedName);
-  expect(id?.args).toEqual(['123']);
-  expect(ast(r)).toMatchInlineSnapshot(`
-    {
-      "table": "issue",
-      "where": {
-        "left": {
-          "name": "id",
-          "type": "column",
-        },
-        "op": "=",
-        "right": {
-          "type": "literal",
-          "value": "123",
-        },
-        "type": "simple",
-      },
-    }
-  `);
+  expect(() => validated(1)).toThrowErrorMatchingInlineSnapshot(
+    `[Error: id must be a string]`,
+  );
+});
 
-  // see comment on `r.hash()`
-  expect(r.hash()).not.toEqual(hashOfNameAndArgs('issue', ['123']));
-  expect(r.hash()).toEqual(
-    hashOfAST((r as StaticQuery<typeof schema, 'issue'>).ast),
+test('defining a synced query with validation and context', () => {
+  const def = syncedQueryWithContext(
+    'myQuery',
+    (id: unknown, createdAt: unknown) => {
+      assert(typeof id === 'string', 'id must be a string');
+      assert(typeof createdAt === 'number', 'createdAt must be a number');
+      return [id, createdAt] as const;
+    },
+    (ctx: object, ownerId: string, createdAt: number) => {
```

Comment:
Can you add a test with the form: `syncedWithContext<string>("foo", validator, fn)` where the signature of fn doesn't need to include types for args or auth.